File: CSharp/Applications/AvaloniaNlogNoDI/Extensions/LoggerExtension.cs
using AvaloniaNlogNoDI.DataStores;
using LogViewer.Core;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using NLog.Target.LogView.Core.Extensions;
using System;

namespace AvaloniaNlogNoDI.Extensions;

public static class ServicesExtension
{
    public static ILoggingBuilder AddNLogTargetsNoDI(this ILoggingBuilder builder, IConfiguration config)
    {
        // We need to use a shared instance of the DataStore to pass to the LogViewerControl
        builder.Services.AddSingleton(MainControlsDataStore.DataStore);

        // call core NLog ServiceExtension initializer
        builder.AddNLogTargets(config);

        return builder;
    }

    public static ILoggingBuilder AddNLogTargetsNoDI(this ILoggingBuilder builder, IConfiguration config, Action<DataStoreLoggerConfiguration> configure)
    {
        builder.AddNLogTargetsNoDI(config);
        builder.Services.Configure(configure);
        return builder;
    }
}

File: VB/Applications/WinFormsLoggingVBNoDI/Extensions/LoggerExtension.vb
Imports System.Runtime.CompilerServices
Imports LogViewerVB.Core
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.DependencyInjection.Extensions
Imports Microsoft.Extensions.Logging
Imports Microsoft.Extensions.Logging.Configuration
Imports MsLoggerVB.Core
Imports WinFormsLoggingVBNoDI.DataStores

Public Module LoggerExtension

    <Extension>
    Public Function AddDataStoreLogger(builder As ILoggingBuilder) As ILoggingBuilder

        builder.AddConfiguration()

        ' We need to use a shared instance of the DataStore to pass to the LogViewerControl
        builder.Services.AddSingleton(MainControlsDataStore.DataStore)
        builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton(Of ILoggerProvider, DataStoreLoggerProvider))

        Return builder

    End Function

    <Extension>
    Public Function AddDataStoreLogger(builder As ILoggingBuilder, configure As Action(Of DataStoreLoggerConfiguration)) As ILoggingBuilder

        builder.AddDataStoreLogger()
        builder.Services.Configure(configure)

        Return builder

    End Function

End Module